packages/components/table/src/store/watcher.ts

```diff
@@ -259,54 +276,53 @@ function useWatcher<T>() {
       return
     }
 
-    let selectedMap
-    if (rowKey.value) {
-      selectedMap = getKeysMap(selection.value, rowKey.value)
-    }
-    const isSelected = function (row) {
+    const { childrenColumnName } = instance.store.states
+    const selectedMap = rowKey.value
+      ? getKeysMap(selection.value, rowKey.value)
+      : undefined
+
+    let rowIndex = 0
+    let selectedCount = 0
+
+    const isSelected = (row: DefaultRow) => {
       if (selectedMap) {
         return !!selectedMap[getRowIdentity(row, rowKey.value)]
       } else {
         return selection.value.includes(row)
       }
     }
-    let isAllSelected_ = true
-    let selectedCount = 0
-    let childrenCount = 0
-    for (let i = 0, j = (data.value || []).length; i < j; i++) {
-      const keyProp = instance?.store?.states?.rowKey.value
-      const rowIndex = i + childrenCount
-      const item = data.value[i]
-      const isRowSelectable =
-        selectable.value && selectable.value.call(null, item, rowIndex)
-      if (!isSelected(item)) {
-        if (!selectable.value || isRowSelectable) {
-          isAllSelected_ = false
+    const checkSelectedStatus = (data: DefaultRow[]) => {
+      let status = true
+
+      for (const row of data) {
+        const isRowSelectable =
+          selectable.value && selectable.value.call(null, row, rowIndex)
+
+        if (!isSelected(row)) {
+          if (!selectable.value || isRowSelectable) {
+            status = false
+            break
+          }
+        } else {
+          selectedCount++
+        }
+        rowIndex++
+
+        if (
+          row[childrenColumnName.value]?.length &&
+          !checkSelectedStatus(row[childrenColumnName.value])
+        ) {
+          status = false
           break
         }
-      } else {
-        selectedCount++
       }
-      childrenCount += getChildrenCount(getRowIdentity(item, keyProp))
+
+      return status
     }
 
-    if (selectedCount === 0) isAllSelected_ = false
-    isAllSelected.value = isAllSelected_
-  }
+    const _isAllSelected = checkSelectedStatus(data.value || [])
 
-  // gets the number of all child nodes by rowKey
-  const getChildrenCount = (rowKey: string) => {
-    if (!instance || !instance.store) return 0
-    const { treeData } = instance.store.states
-    let count = 0
-    const children = treeData.value[rowKey]?.children
-    if (children) {
-      count += children.length
-      children.forEach((childKey) => {
-        count += getChildrenCount(childKey)
-      })
-    }
-    return count
+    isAllSelected.value = selectedCount === 0 ? false : _isAllSelected
```

Comment:
```suggestion
    isAllSelected.value = selectedCount === 0 ? false : checkSelectedStatus(data.value || [])
```

Reply:
The checkSelectedStatus function needs to be executed here first, so it cannot be modified like this

Reply:
ok.